[spotify] Spotify Binding initial contribution

astenlund74 opened the pull request:
Here's an implementation of a binding to control Spotify Connect devices through the Spotify Web API. It is far from completed and is a to a large extent quick 'n dirty implementation. I have tried to clean it up a bit. There are several documented TODOs and many more unkown. :-)

The Spotify authorization flow for the Web API is not a perfect fit for OpenHAB bridge initiation. It requires users to go to a separate web url to initiate the Spotify login and associate auth token with the bridge.

The whole process is described in the inital README.md file.

Please, review and come with feedback.

Signed-off-by: Andreas Stenlund <andreas@stenlund.net> (github:
astenlund74)

kubawolanin commented:
Hey @astenlund74 thanks for your contribution!

I remember @tavalin working on [Spotify-connect](https://github.com/tavalin/openhab2-addons/tree/spotify_co/addons/binding/org.openhab.binding.spotifyconnect) binding a while ago.
You may join your forces on that one :-)

astenlund74 commented:
@kubawolanin,talk about finding a needle in a haystack ... had I only known about it!  :-) I have been searching like crazy for implementations and read some forum posts asking for this kind of binding, but no reply and I decided to give it a go.

I am new to this community work and contributions, should I realign my implementation and incorporate it into what he has started?

Looking at the implementation @tavalin has there, it seem to be kind of very early stage, maybe there is more code around somewhere? 

This contribution, even if _far from completed_, is a full-blown working binding.

kubawolanin commented:
Funny thing is I've just stumbled upon Daniel's binding idea on the [community forum](https://community.openhab.org/t/idea-for-binding-spotify-connect-binding/11290/6?u=kubawolanin)

> should I realign my implementation and incorporate it into what he has started?

No, I think it's good to continue discussion there, in this PR. My proposal was rather to @tavalin, whether he'd like to share his thoughts on the topic, since he was involved in it before.

Keep up the good work! Hopefully I'll find some time to test your binding on the weekend :)

BTW. Your binding is not visible on the [jfrog builds list](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/) - could you attach compliled JAR here?
Thanks!

tavalin commented:
Hi

To echo what @kubawolanin said, thanks for your contribution @astenlund74.

As mentioned my work was an early investigation (which unfortunately has been on the backburner).  I was using the [spotify-http](https://github.com/Spotifyd/spotifyd-http) server to do all the heavy lifting for the Spotify Connect communication as Spotify's Web API didn't seem to allow you to discover and control Connect devices - I'm not sure if that has changed in the last 8 months or so.

Hopefully I will have some time this week to test and provide my thoughts.

martinvw commented:
Thanks for your PR! 

Please include the build in the parent `pom.xml` in that case it will already be tested and analysed by travis.

Please also check:

http://docs.openhab.org/developers/development/bindings.html
http://docs.openhab.org/developers/development/guidelines.html

astenlund74 commented:
I found and removed some strict versions on imports in the MANIFEST which caused issues when testing on my PI3 "production" installation (running release 2.0 build from repositories). 

I have also added org.openhab.binding.spotify to the feature and binding pom.xml files. I thought I did that, but must have messed up with git somewhere... hence the unsigned merge above I think, how to fix?

Travis is building and according logs the binding has been processed now. Hopefully everything looks good enough. 

I added the binary jar in its own repo here for you to test: https://github.com/astenlund74/org.openhab.binding.spotify-binary

@tavalin, all feedback and help is welcome and appreciated. I think Spotify provided the new Web API endpoints quit recently - they are still BETA which might be something to consider when running this binding. So you probably started your work too early - I took the easy path. :-)

I haven't noticed any considerable issues with the API other that the normal Spotify intermittent failures to detect Spotify Connect devices on the network - the same issues I see using Android app.

tavalin commented:
Hi @astenlund74 

I compiled your PR last night but ran out of time to do any proper testing.  Some initial comments:

- Your README was helpful for getting setup but there were a few typos and formatting issues.  Also, I think most people probably use a browser on a different machine to their OH server so the mention of localhost may confuse someone new.

- Authorizing openHAB to use my Spotify account worked. 👍 

- Discovery of my Spotify Connect devices also worked well. 👍 

- When inspecting a Spotify Connect thing it appears to have lots of channels - most of them I would probably mark as "advanced" e.g. trackUri, refreshToken, etc.

- I used PaperUI to create an item which I linked to the artist channel but I didn't see any updates in my log, even after changing tracks.

- I noticed when having a VERY quick look through the code that the URIs aren't defined as constants so I assume a new version gets called everytime you call those control methods.

A good start so far...will try and do a thorough GitHub review once I get more time.

tavalin commented on addons/binding/org.openhab.binding.spotify/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,308 @@
+
+                                                 
+    <!-- Spotify Player Device -->
+    <bridge-type id="player">
```
Is this Thing essentially the account details/tokens?

tavalin commented on addons/binding/org.openhab.binding.spotify/README.md:
```diff
@@ -0,0 +1,114 @@
+
+The binding requires you to register an Application with Spotify Web API on https://developer.spotify.com 
+
+When registering your new Spotify Application for OpenHAB Spotify Bridge you have to specify the allowed "Redirect URIs" aka white-listed addresses. Here you have to specify the address to you Eclipse Smarthome server. 
```
Should be "openHAB"

tavalin commented on addons/binding/org.openhab.binding.spotify/README.md:
```diff
@@ -0,0 +1,114 @@
+If you run it on the local machine with default settings you add "http://localhost:8080/connectspotify/authorize" to the Redirect URIs. The authentication process takes place using your client web browser so this Redirect URI is where the OpenHAB web UI is.
+
+Steps required to get this binding running:
+0. Install bind and make sure the Spotify Bridge things are available on your server
```
*binding

tavalin commented on addons/binding/org.openhab.binding.spotify/README.md:
```diff
@@ -0,0 +1,114 @@
+Steps required to get this binding running:
+0. Install bind and make sure the Spotify Bridge things are available on your server
+1. Register on https://developer.spotify.com with your Spotify Premium Account
+2. Add a new Application. Add Redirect URI for OpenHAB. We need the "Client ID" and "Client Secret" to add a new binding.
```
*openHAB

tavalin replied:
Initially I missed the "Add Redirect URI for openHAB" bit as it didn't follow the screen flow.  Maybe move that sentence until after the user has input the Client ID and Client secret.

astenlund74 replied:
Many thanks for all your review and feedback! 

I will take care of that as soon as possible. I already started changes on the channels and categorized advanced ones. I have an updated README ready as well. I also  noticed some polling thread issue which I didn't  manage to solve and push yesterday. Won't have time until Sunday evening to work on this again.

tavalin commented on addons/binding/org.openhab.binding.spotify/README.md:
```diff
@@ -0,0 +1,114 @@
+
+When registering your new Spotify Application for OpenHAB Spotify Bridge you have to specify the allowed "Redirect URIs" aka white-listed addresses. Here you have to specify the address to you Eclipse Smarthome server. 
+
+If you run it on the local machine with default settings you add "http://localhost:8080/connectspotify/authorize" to the Redirect URIs. The authentication process takes place using your client web browser so this Redirect URI is where the OpenHAB web UI is.
```
*openHAB

tavalin commented:
@astenlund74 I think the Thing modelling needs to be revised.  

In your implementation each Spotify Connect Thing has it's own playback controls but that doesn't reflect the reality of how the Spotify API works. i.e. I can't have two players linked to my account playing music at the same time (outside of the occasional glitch).  Additionally if music is playing on my PC I can use my phone to to pause/play/skip(or even take over) the music and therefore the player control is account wide.

I'm thinking we could probably simplify it a bit:

1. Have one Thing that represents your Spotify Account that has channels for controlling the music, displaying the title, artist, album etc. 
2a. Have a channel that represents the Spotify Connect Player to play the music..  The possible options may be dynamic based on the /player/devices results.

OR

2b. Have a Thing for each Spotify Connect Player discovered that has a switch channel to takeover playback.

tavalin commented:
This screenshot shows I think option 1 could work.  I would however have Spotify Connect Device Name as a selection drop down dynamically created from the available Spotify Connect devices returned from the last poll.

![image](https://cloud.githubusercontent.com/assets/1771518/26515043/94e0a46a-426e-11e7-8214-a33cdfa86326.png)

tavalin commented:
@astenlund74 did you get a chance to look at my comments?

tavalin commented:
@astenlund74 any chance to look at this yet?

astenlund74 commented:
@tavalin, sorry for delay in response and updates. Busy days and long running test of the Web API polling thread would be my excuse if I need one! :-)

I do understand your thoughts regarding the model, I have given the model some thoughts too. I did reconsider it during this time of silence and I have run additional Web API tests, but I am not convinced that the model really requires modification. Let me present my view on the current implementation. 

I see a difference in the models of the Web API and the underlying Spotify platform as seen in the apps. A dedicated connect device hardware is different from an app e.g. on a smartphone. An app is always in the context of the logged in user account and you usually do not log out/in with different accounts on your smartphone even if you could. The connect device hardware seem to always be left associated with the user account that played on it most recently. This can be overridden from a smartphone or computer app when you are on the same network as the connect device hardware (local discovery & override or local discover & central override?). The Web API does not allow to take over playing on a device from another user as the control is from Spotify servers and not directly on the local network (unless it is only restricted). I think this is a matter of security as there is no user authentication per device. Maybe this will be allowed through Web API in the future through use of separate device authorization tokens or similar.

However for all devices (apps and connect hardware) currently in the context of your user account, the Web API allows you to use the player controls for a particular device; the play/pause/next/prev-calls all accepts an optional device parameter. In that sense all devices allows for player control. If the device is not active it will become the currently active playing device.

The bridge currently acts in the context of a specific user account and the binding allows to add multiple bridges with different account tied to them. In that sense I believe the model is consistent.

The Player control channel on the bridge thing is the common user player control whereas the Player control channel on device things are dedicated to the device. If you start playing on an inactive device you will transfer playback to that particular device. If someone else is playing on the device it will simply not be online.

In wanted the binding to be flexible and allow for several use cases, but still keep it as simple as possible. Removing devices and having only a dynamic device channel on the player to select device could be an option, but that would imply the app functionality of being able to take over control of the device which is not possible.

Having user accounts - players - as separate things under the bridge would require an authorization process to be completed for each player thing. I think that would just add more confusion in the process of configuring the binding. If you have more than one account tied to the bridge, which should then be used in discovery of devices? Which account is used for playback on device channels? I don't think it is an easily understood model.

It makes sense to me that the bridge represents the user account context and also the primary Player control. It is convenient to have the Player control on the devices allowing for a straight forward configuration of UIs for your "favorite living room player".

With the current model the same device can exist multiple times as a thing which might cause a bit of confusion, but they will always exist in the context of the user account associated with it's bridge and the Thing UID also has the bridge id in it to emphasize that. It is not possible to move the thing to another bridge in the current model.

One potential (complex) model change, which I am not sure is possible with the framework, would be to have connect devices as unique things that could be automatically associated with the bridge of the user account it is currently in scope of. That would allow you to discover devices on multiple bridges/user accounts and increase possibility of having them online independent of which user account had them active most recently. 

With all that said, I still think your proposal of an advanced dynamic channel on the bridge  allowing to set active device is a good one and I will proceed with that.

tavalin commented:
@astenlund74 I haven't had chance to test your latest commits but I just wanted to check that you weren't waiting for me? I wouldn't want this to stall as I think a Spotify Connect binding would be a great addition.

tavalin commented:
@astenlund74 I haven't had chance to test your latest commits but I just wanted to check that you weren't waiting for me? I wouldn't want this to stall as I think a Spotify Connect binding would be a great addition.

astenlund74 commented:
@tavalin No, thanks for waking me up! 👍 I didn't really know what to do next in this process and I also intended to add the dynamic channel functionality on bridge as you proposed but has been busy with other holiday activities. I've got the latest binding running at home for a longer period now and it is robust here - even with a couple of weeks on pause the playing kicks in right away when activating so the active session management seem to work well. Mentioning that, one enhancement could be not to renew session tokens when API is paused for a longer period of time and instead logon again at next call to the API. This would make less use of resources.
If the contribution is considered GO, let's proceed with merge.

tavalin commented:
@astenlund74 it seems as if there are still some code compilation and code quality issues in your PR.  Could you have a look please?

astenlund74 commented:
@tavalin yes, I am looking into it. will submit updates shortly.

felixloesch commented:
Cool stuff. Will the binding also allow start of playback with a playlist  a on a given  A Spotify connect device? This would be awesome to start Spotify playback in OH rules.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+    private int tokenValidity = 3600;
+
+    private SpotifyHandler spotifyPlayer = null;
+    private final ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
```
Do we really need 3 extra thread pools?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+    private final ExecutorService syncExecutorService = Executors.newFixedThreadPool(1);
+    private final ExecutorService asyncExecutorService = Executors.newFixedThreadPool(1);
+
+    @SuppressWarnings({ "rawtypes" })
```
Why don't you resolve this warning, or can it not be resolved?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+            } catch (Exception e) {
+                logger.error("Error starting HttpClient", e);
+            }
+
```
Loose some newline here and there at the beginning / end of blocks.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+        scheduledExecutorService.shutdownNow();
+        syncExecutorService.shutdownNow();
+
+        // TODO: Not really need? Causing bad exceptions when refreshing SpotifySession in Controller check later....
```
There is a TODO, what should be done and can it be completed before merging the PR?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+        }
+
+        scheduledExecutorService.shutdownNow();
+        syncExecutorService.shutdownNow();
```
Why do you leave one open? I.e. you declared three in this class.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+    public static SpotifySession getInstance(SpotifyHandler player, String clientId, String clientSecret,
+            String refreshToken) {
+
+        SpotifySession session = new SpotifySession(player, clientId, clientSecret, refreshToken);
```
Just return it

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+
+                if (response.getStatus() == 200) {
+                    Gson gson = new Gson();
+                    SpotifyWebAPIAuthResult test = gson.fromJson(response.getContentAsString(),
```
`test` does not look like a proper name :-)

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+                logger.trace("Response Data: {}", response.getContentAsString());
+
+                if (response.getStatus() == 200) {
+                    Gson gson = new Gson();
```
This `Gson` be reused and is a little bit expensive to create, so please reuse

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+
+        try {
+            refreshToken();
+        } catch (java.lang.NullPointerException npe) {
```
Better make sure you don't get that please work on that.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+            return false;
+        }
+
+        // TODO: Find a more suitable to retrieve token validity if it changes after being scheduled? Scheduling refresh
```
There is a TODO, what should be done and can it be completed before merging the PR?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+        for (int i = 0; i < HTTP_CLIENT_RETRY_COUNT; i++) {
+            logger.debug("Spotify Web API call attempt: {}", i + 1);
+            try {
+                String url = "https://accounts.spotify.com/api/token";
```
This looks familiar can you apply some re-use?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+ *
+ * @author Andreas Stenlund - Initial contribution
+ */
+public class SpotifySession implements Runnable {
```
I think its not a proper design if the `SpotifySession` is a `Runnable` containing 3 additional executors, can you please refactor that?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+
+        try {
+            while (!future.isDone()) {
+                Thread.sleep(ASYNC_RESULT_WAIT_TIME);
```
Why like this? You are still stalling your thread although you execute the work in another?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+            }
+            return future.get();
+
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+                logger.debug("Spotify Web API call attempt: {}", i + 1);
+                try {
+                    response = httpClient.newRequest(url).method(HttpMethod.fromString(method))
+                            .header("Authorization", "Bearer " + accessToken).header("Accept", "application/json")
```
This looks familiar can we apply some re-use

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+                                "Spotify Web API returned code 429 (rate limit exceeded). Retry After {} seconds. Decrease polling interval of bridge! Going to sleep...",
+                                retryAfter);
+
+                        Thread.sleep(Integer.parseInt(retryAfter));
```
Thead sleeps is most of the time not the solution, the problem we are trying to resolve is that we should processing the queue for a certain time? Maybe a different setup is needed for that?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+    public class SpotifyWebAPIAuthResult {
+
+        @SerializedName("access_token")
+        @Expose
```
Is Expose needed?

martinvw replied:
Check others.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+
+        }
+
+        public class Device {
```
All classes should have a description and author header and given the size of them a separate package to contain all the Spotify objects seems sensible.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifySession.java:
```diff
@@ -0,0 +1,1299 @@
+     */
+    public class SpotifyWebAPIDeviceList {
+
+        public class Device {
```
Do these classes need to be inner classes of each other or can they just be static inner classes? And thus also just separate classes in there own files.

data219 commented:
@astenlund74 Are you still working on this binding?

fuslwusl commented:
The Binding works great. Why is it still in status pending?

martinvw commented:
@astenlund74 ping?

fuslwusl commented:
5 month without response from @astenlund74 is not a good sign. Can anyone else make the required changes to resolve those conflicts? It's a really great and powerful binding and it would be crazy to let it die in this state.

fuslwusl commented:
I am not a developer but I can help testing and give you feedback if you like.

astenlund74 commented:
@martinvw & @tavalin - sorry for the absence! 
Also sorry that I just lost all inspiration to continue the work on this binding with the last review. I just wanted to share my work for the best of the community - the binding does what it is supposed to do and works fine for me and fulfill my requirements.
@mgbowman is willing to continue development and I have added him as collaborator to my repository for the PR. Hopefully the work can be finalized by him to meet your requirements and allow the PR to be approved and submitted. I will support him in this.

mgbowman commented:
@martinvw how do you want me to proceed? Do you prefer I just commit directly to @astenlund74's repo and try and get this PR merged?

martinvw commented:
Great to see such an arrangement, @astenlund74 thank you very much for the initial work and your support for the further development.

@mgbowman thanks for taking it over.

You should be able to add a second remote (that is a git term) to your existing dev environment, if you check out astenlund74's master under different local name you should be fine.

mgbowman commented:
@martinvw I figured it out! 

```
git remote add astenlund74 https://github.com/astenlund74/openhab2-addons.git
git fetch astenlund74
git checkout -b spotify-binding astenlund74/master
```

Sometimes it feels like it will take a lifetime to properly learn git 😆

fuslwusl commented:
Great to see that work on this binding is continued!  Thanx for this great binding - i'll test it when there is a new version available!

Hilbrand commented:
@mgbowman Do you need  some help finishing this binding?

mgbowman commented:
@Hilbrand to be honest, I've only ever looked through the code - I haven't had time to work on this as I'm super busy working on the electrical / automation of my remodel. If you want to give it a go, by all means go wild!

My plan of attack was to actually refactor the binding to make use of this 3rd party Spotify Java lib:

https://github.com/thelinmichael/spotify-web-api-java

It appears to implement everything needed by the current binding and it would hopefully make development easier for others moving forward.

martinvw commented:
> @mgbowman Do you need some help finishing this binding?

Great job @Hilbrand 

@mgbowman good luck with you remodeling

Hilbrand commented:
@astenlund74 can you add me as collaborator to your repository? I see you have used master branch for this pull request. This means updates to your repository need to done on master. If you are not going to develop anything in the near future this should not directly be a problem. Because the branch requires rewritting and rebasing and then if you would  also want to use it you need to make sure you rebase correctly and not work on master unless it's related to this or.

astenlund74 commented:
Hi

You're welcome to help out. I guess some of the issues I have had with git
is related to the master and not using a separate branch for the changes.

Best regards,
Andreas

Den ons 25 juli 2018 kl 14:32 skrev Hilbrand Bouwkamp <
notifications@github.com>:

> @astenlund74 <https://github.com/astenlund74> can you add me as
> collaborator to your repository? I see you have used master branch for this
> pull request. This means updates to your repository need to done on master.
> If you are not going to develop anything in the near future this should not
> directly be a problem. Because the branch requires rewritting and rebasing
> and then if you would also want to use it you need to make sure you rebase
> correctly and not work on master unless it's related to this or.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/2295#issuecomment-407738360>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AXPtxpUUv5Be9_jBWEMYqewGEW7LMTP_ks5uKGVLgaJpZM4NkM7x>
> .
>

fuslwusl commented:
Hi, is there a new jar file of the binding for testing? Has anybody fixed the problems of the binding?

Hilbrand commented:
@fuslwusl No not yet. I'm working on it. I'll post an update in the community Spotify binding thread when it's available.

Hilbrand commented:
I've push forced my finishing of this binding. There are a lot of changes, expect on the things interface. See my commit message for an overview of all the changes.

Hilbrand commented:
Travis fails on a missing requirement. I've added a Dashboard Tile and need to import the module. But I have no idea what would be the correct way to do it. I tried adding a dependency to the feature.xml, but that didn't work. So any ideas are welcome.

Here is the travis error:
> ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: org.openhab.binding.spotify 2.4.0.qualifier
[ERROR]   Missing requirement: org.openhab.binding.spotify 2.4.0.qualifier requires 'package org.openhab.ui.dashboard 0.0.0' but it could not be found

kaikreuzer commented:
>  I've added a Dashboard Tile and need to import the module

Hm, bindings are not supposed to add dashboard tiles, thus this dependency is illegal. Could it be simply removed...?

Hilbrand commented:
Hmm... It looks so cool :disappointed: This is to get an easy access for the user to the local page to authorize the Spotify account. If there is another way to get the link to this page to the user then I'd happy to know how. Btw. Why shouldn't bindings add dashboard tiles?
![csp](https://user-images.githubusercontent.com/1272873/45506967-18dbd080-b791-11e8-9f29-4382f8939d46.png)

kaikreuzer commented:
> This is to get an easy access for the user to the local page to authorize the Spotify account.

I assume this is for an OAuth client flow. I would actually like to see support for this built into the Paper UI, so that such a redirect is automatically done when setting up a bridge/thing in the Paper UI. Currently, that's not yet available, so the easiest option might be to add a link to the config parameter description of the token.

> Why shouldn't bindings add dashboard tiles?

Because it would be a nightmare if every binding does that. And the tiles should be an entry point for the user. Authorising an account is rather a step somewhere in a user flow, which should not be isolated here.

kaikreuzer commented:
The best example of how to get hold of OAuth tokens is https://www.openhab.org/addons/bindings/innogysmarthome/#thing-configuration.

It is currently half-automated - in the Paper UI, you can click the link to the auth page and you will be redirected to a page that displays the token, which you can then manually copy&paste into your config.

I'd suggest to follow the same approach.

Yes, openhab-cloud could be useful for all OAuth app authorisation, but that's a whole different story. There is a related issue for that already: https://github.com/openhab/openhab-cloud/issues/133

Hilbrand commented:
The Innogy approach uses hard-coded clientId and clientSecret values. For the Spotify binding this is not possible. But I can extend the myopenhab cloud page to let the user enter a clientId and then construct a url that will be directed to Spotify. Code-wise it will be simpler as there is no need for the servlet part. But the user needs to do a little more; the user needs to do an extra step: enter the authCode, something that is currently done automatically. I'll see what I can make of it.

Interestingly the Innogy binding makes use of the google oauth library, but uses the built-in apache http transport, and not the smart home provided http service (Which would require additional work).
I also looked at the Smart Home oauth2 implementation. However it looks less complete as the google implementation and doesn't have the nice api like the Google library has. Also to use it with the Spotify binding would required additional work, for example it doesn't handle all possible http statuses, which would cause the Spotify binding to report errors. These kind of things make me reluctant to use the Smart Home Oauth implementation, because I'm worried I'll be debugging the library instead of finishing this binding. Personally I would prefer an implementation that uses/looks like the google library, but I don't know if that is an option/direction/stragey to go to. (For example like with serial to have 2 different serial module implementations)

kaikreuzer commented:
> But I can extend the myopenhab cloud page to let the user enter a clientId and then construct a url that will be directed to Spotify.

You mean the static page on www.openhab.org (note that the binding must not have a dependency on myopenHAB)?

> Interestingly the Innogy binding makes use of the google oauth library, but uses the built-in apache http transport, and not the smart home provided http service

Only because it has been developed and merged before the work on the ESH OAuth client library had started. The plan was to replace the used implementation sooner or later.

> However it looks less complete as the google implementation and doesn't have the nice api like the Google library has.

I thought it should be rather simple to use. If you have suggestions on how the API can or should be improved, please file according issues at ESH. Certainly this API will only meet all needs, if it evolves and incorporates feedback, but for this we need an active participation on it.

> but I don't know if that is an option/direction/stragey to go to

Well, it would mean to revert the current decision to have our own API.
So if you'd succeed to use the Google lib in a lightweight way (by plugging in your own http transport, which sounds to me like quite some error-prone job...) and have strong points why the API is so much better, the discussion can certainly be re-opened.

Hilbrand commented:
I've finished the work on this binding. So it can be reviewed again. The latest changes, I've pushed, integrate this binding to use the built-in OAuth2 support in Eclipse Smart Home. I waited until the official release of 2.4.0 because there where some changes in Eclipse Smart Home related to this binding. With these changes in an official  release it's now possible to run this binding on a stable openHAB release.

So I'm going to ping you all here :smile: @tavalin and @martinvw you both did a review before. If you would be so kind and have a second look I'd be grateful. @mgbowman you did some work before I took over, and @astenlund74, as the originator of this binding, if you're still attached and want to add something please do so.

rkantos commented:
I tested out this and seems to work fine so far. The only thing I found confusing was the play/pause/next/prev commands, which I could only get working with habpanel buttons when having the commands in all caps(PLAY/PAUSE..)!

martinvw commented on addons/binding/org.openhab.binding.spotify/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,366 @@
+
+			<channel id="deviceId" typeId="activeDeviceId" />
+			<channel id="deviceName" typeId="activeDeviceName" />
+			<channel id="deviceType" typeId="activeDeviceType" />
```
Should this be a channel or a property same for the two above?

martinvw commented on addons/binding/org.openhab.binding.spotify/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,366 @@
+	</channel-type>
+	<channel-type id="currentlyPlayedTrackHref" advanced="true">
+		<item-type>String</item-type>
+		<label>Track href</label>
```
This label is not very descriptive, it also does not follow the capitalised-words syntax.

martinvw commented on addons/binding/org.openhab.binding.spotify/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,366 @@
+	</channel-type>
+	<channel-type id="currentlyPlayedTrackId" advanced="true">
+		<item-type>String</item-type>
+		<label>Track Id</label>
```
What kind of id is this, is it useful?

martinvw replied:
Okay

martinvw commented on addons/binding/org.openhab.binding.spotify/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,366 @@
+
+	<channel-type id="currentlyPlayedAlbumId" advanced="true">
+		<item-type>String</item-type>
+		<label>Album href</label>
```
This label is a duplicate from the next ~~one~~ *two* and is not capitalised like the rest.

martinvw commented on addons/binding/org.openhab.binding.spotify/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,366 @@
+	</channel-type>
+	<channel-type id="currentlyPlayedArtistHref" advanced="true">
+		<item-type>String</item-type>
+		<label>Artist href</label>
```
Please check all labels

martinvw commented on addons/binding/org.openhab.binding.spotify/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,37 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-SymbolicName: org.openhab.binding.spotify;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.5.0.qualifier
```
👍

martinvw commented on addons/binding/org.openhab.binding.spotify/README.md:
```diff
@@ -0,0 +1,225 @@
+| refreshPeriod | This is the frequency of the polling requests to the Spotify Connect Web API in seconds.                                                                       |
+
+The following configuration option is available on the Spotify device:
+| Parameter | Description                                           |
```
I think that there should also be a newline before a table, please check the SAT findings.

martinvw replied:
O, I might be wrong then 🤔

martinvw commented on addons/binding/org.openhab.binding.spotify/pom.xml:
```diff
@@ -0,0 +1,18 @@
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>org.openhab.binding</groupId>
```
This looks missformatted (mix of tabs and spaces)

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAccountHandler.java:
```diff
@@ -0,0 +1,79 @@
+     * @param reqCode The unique code passed by Spotify to obtain the refresh and access tokens
+     * @return returns the name of the Spotify user that is authorized
+     */
+    String authorize(String redirectUrl, String reqCode);
```
This method is also public, can you leave out the other `public`s

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAuthService.java:
```diff
@@ -0,0 +1,176 @@
+    private @Nullable SpotifyAccountHandler getSpotifyAuthListener(String thingUID) {
+        final Optional<SpotifyAccountHandler> listener = handlers.stream().filter(l -> l.equalsThingUID(thingUID))
+                .findFirst();
+        return listener.isPresent() ? listener.get() : null;
```
I think this can be written differently `listener. orElse(null)` Consider naming optional things more explicit, what about a `maybeListener`, you will be more likely to check before calling `.get()` on it.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAuthServlet.java:
```diff
@@ -0,0 +1,202 @@
+                try {
+                    replaceMap.put(KEY_AUTHORIZED_USER, String.format(HTML_USER_AUTHORIZED,
+                            spotifyAuthService.authorize(servletBaseURL, reqState, reqCode)));
+                } catch (RuntimeException e) {
```
Do we  really need to catch this broad?

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/SpotifyAuthServlet.java:
```diff
@@ -0,0 +1,202 @@
+            try {
+                final String key = m.group(1);
+                m.appendReplacement(sb, Matcher.quoteReplacement(map.getOrDefault(key, "${" + key + '}')));
+            } catch (RuntimeException e) {
```
Do we have to catch this broad especially here I would expect that the range of exceptions is small

martinvw replied:
Did you address this one?

Hilbrand replied:
Same comment as above. It's in a servlet. This just catches all unchecked exceptions, like NullPointers. Normally those are not catched as it's unknown what is thrown unlike checked exceptions. So I would have to guess what exceptions would be thrown. This is just very defensive programming to avoid showing mangled pages to the user whatever happens.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/api/SpotifyApi.java:
```diff
@@ -0,0 +1,293 @@
+        final String url = "play" + optionalDeviceId(deviceId, QSM);
+        final String play;
+        if (trackId.contains(":track:")) {
+            final String jsonRequest = "{\"uris\":[%s],\"offset\":{\"position\":0}}";
```
😕

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/api/SpotifyApi.java:
```diff
@@ -0,0 +1,293 @@
+            }
+        } catch (IOException e) {
+            logger.debug("Request failed to during refresh token: ", e);
+            throw new SpotifyException(e.getMessage());
```
Consider to not discard the the stacktrace here

Hilbrand replied:
I've added exception as a parameter to the SpotifyException class and replaced all with the exception as argument and removed the logging. I've done this for all cases where it applies.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/api/SpotifyApi.java:
```diff
@@ -0,0 +1,293 @@
+            throw new SpotifyException(e.getMessage());
+        } catch (OAuthException | OAuthResponseException e) {
+            logger.debug("Request authorization failed to during refresh token: ", e);
+            throw new SpotifyAuthorizationException(e.getMessage());
```
Consider to not discard the the stacktrace here

martinvw replied:
As far as I know its best practice to either log or throw, but not you are kinda doing both, I think that passing the exception as a cause to the new one should eliminate the logging here

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/api/SpotifyApi.java:
```diff
@@ -0,0 +1,293 @@
+        final Function<HttpClient, Request> call = httpClient -> httpClient.newRequest(url).method(method)
+                .header("Accept", CONTENT_TYPE).content(new StringContentProvider(requestData), CONTENT_TYPE);
+        try {
+            try {
```
I think this is not ideal, would it make sense to split up the methods so that we do not have to explicitly nest two try-statements.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyBridgeHandler.java:
```diff
@@ -0,0 +1,605 @@
+        } catch (RuntimeException | OAuthException | IOException e) {
+            logger.debug("Authorize failed: ", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            throw new SpotifyException(e.getMessage());
```
Should you update the status, log and rethrow? Please apply passing the cause here and not logging when rethrowing.

Hilbrand replied:
I've removed the logging here. It still should update the status. This method is called from the servlet to pass authorization. If that failed this is reflected in the status so the user knows what is wrong. Because the error message itself is only known here the status must be updated here.

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyBridgeHandler.java:
```diff
@@ -0,0 +1,605 @@
+            throw new SpotifyException(e.getMessage());
+        } catch (OAuthResponseException e) {
+            logger.debug("Authorize failed: ", e);
+            throw new SpotifyAuthorizationException(e.getMessage());
```
See above

martinvw commented on addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyBridgeHandler.java:
```diff
@@ -0,0 +1,605 @@
+     * @return true if method completed without errors.
+     */
+    private boolean pollStatus() {
+        synchronized (scheduler) {
```
What is the intention here, you share the `scheduler` object with a lot of other things, AFAIK you should share on class-local things. What about creating a separate `Object` instance to synchronise over, now it looks kinda of incorrect because a lot of usages of `scheduler` or not synced.

martinvw replied:
Did you address this one?

martinvw replied:
Never mind my phone confused me

Hilbrand replied:
Introduced a new field that is used instead of the `scheduler` object.

Hilbrand commented:
BTW what's wrong with the jenkins build? I looked at the log and looked at the smart home snapshot urls. But all Eclipse Smart Home urls (https://ci.openhab.org/artifactory/libs-snapshot/org/eclipse/smarthome/) result in a 404.

martinvw commented:
@kaikreuzer can you take a look at this one?

martinvw commented:
@kaikreuzer I already approved, it would be great if you could take a quick glance, how much can be wrong with @Hilbrand building and me approving 😉

kaikreuzer commented:
> how much can be wrong with @Hilbrand building and me approving

You know what: I merge without having any glance at it. If two @openhab/2-x-add-ons-maintainers have approved it, we can clearly assume that it is ready to be merged, so don't bother about me, but feel free to move things forward yourself! Thanks to all of you for the awesome work!

ehora commented:
Hello,

not sure if this is the right place, but I think there is a problem with the SpotifyDynamicStateDescriptionProvider:

https://github.com/openhab/openhab2-addons/blob/0557788d8fe05a704acc9828591e4ae09d28bcdc/bundles/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyDynamicStateDescriptionProvider.java#L80

I think the "else"-case is wrong. Because now, if the spotify binding registers first, it blocks all other DynamicStateDescriptionProviders from serving their own channels.

Is my unterstanding correct?

Kind regards,
Oliver
